### ISSUE TITLE
Doesn't seem to work in other documents

### ISSUE THREAD

**Issue by LoganDark** [focus-trap/focus-trap-react#539]
I appreciate the effort you have put into making the focus trap library work in iframes and child windows, but unfortunately it seems like maybe the React version is not prepared for this?

I'm using the FocusTrap like `<FocusTrap active={element !== null} focusTrapOptions={{document}}>`, where `document` is the document of the child window (derived from `element.ownerDocument`), but the focus trap still isn't trapping focus. When the same component is used in the main document, it works just fine.

Do you know what could be the cause of the issue?

**Comment by stefcameron** [comment]
You're welcome! 🙂 

It's quite possible there are some things the code isn't handling. I had another look at the code, and I found one case where it's not respecting the `focusTrapOptions.document` setting: https://github.com/focus-trap/focus-trap-react/blob/master/src/focus-trap-react.js#L66

When you say it doesn't work, have you checked your console for errors? If you're hitting the issue above, you might have exceptions in the console from line 68, and that would cause the trap not to function properly.

If not, then something else is going on.

> I'm using the FocusTrap like `<FocusTrap active={element !== null} focusTrapOptions={{document}}>` [...]

Just a thought here, and probably wouldn't change anything, but `{{document}}` in there gives me the impression you're storing the document you want to give to focus-trap into a variable, like `const document = something` which means you're shadowing the global `document`. This probably works, but just because you named it `document`, maybe it's not the `document` you think it is? Shot in the dark...

> [...] where document is the document of the child window (derived from `element.ownerDocument`), [...]

So your code is more like `<FocusTrap active={element !== null} focusTrapOptions={{document: element.ownerDocument}}>`?

Also, what is `element` in `active={element !== null}`? Sounds like an HTML Element obtained through a Ref. It would be better to avoid this if possible and use a state variable, for example, to decide if the trap should be active or not.

Another idea is to put the trap _in_ the other document. For example, if you're using this for a modal window of some kind, put the trap in that modal component, and create it only when that modal is created, instead of keeping it "alive" outside of it and trying to activate it only when the modal is visible.

**Comment by LoganDark** [comment]
> When you say it doesn't work, have you checked your console for errors? If you're hitting the issue above, you might have exceptions in the console from line 68, and that would cause the trap not to function properly.
> 
> If not, then something else is going on.

No errors.



> Just a thought here, and probably wouldn't change anything, but `{{document}}` in there gives me the impression you're storing the document you want to give to focus-trap into a variable, like `const document = something` which means you're shadowing the global `document`. This probably works, but just because you named it `document`, maybe it's not the `document` you think it is? Shot in the dark...

Indeed I am storing it in a variable:

```ts
	const document = useMemo(() => menu?.ownerDocument, [menu])
```

`menu` is a state variable that is updated using callback refs, I can confirm it's set correctly (and `document` too).

> Sounds like an HTML Element obtained through a Ref. It would be better to avoid this if possible and use a state variable, for example, to decide if the trap should be active or not

If it was a ref, it would be `element.current !== null`, which would be incorrect because changing the ref wouldn't update the component.

> Another idea is to put the trap _in_ the other document. For example, if you're using this for a modal window of some kind, put the trap in that modal component, and create it only when that modal is created, instead of keeping it "alive" outside of it and trying to activate it only when the modal is visible.

The trap _is_ in the other document. It wraps a div _whose document is the document I'm passing to focusOptions_:

```ts
	const [menu, setMenu] = useState<HTMLDivElement | null>(null)
	const document = useMemo(() => menu?.ownerDocument, [menu])

	// a small amount of other stuff but you get the point...

	return <FocusTrap active={menu !== null} focusTrapOptions={{document}}>
		<div ref={mergeRefs(setMenu as Ref<HTMLDivElement>, ref)} className={extendClassname(className, 'context-menu')} tabIndex={0}
		     onKeyDown={(e: KeyboardEvent<HTMLDivElement>) => e.stopPropagation()} {...others}>
			{children}
		</div>
	</FocusTrap>
```

**Comment by stefcameron** [comment]
OK, then what do you mean when you say it's "not working"? Is it that the trap activates, and traps focus _within_ the child document, but you can still click on elements _outside_ (in the host document)? That's a limitation of having the trap work on a child document.

Also, just to check everything, you should be using focus-trap-react v8.8.0 or greater. That's the version that added support for the focus-trap `document` option.

**Comment by LoganDark** [comment]
> OK, then what do you mean when you say it's "not working"? Is it that the trap activates, and traps focus _within_ the child document, but you can still click on elements _outside_ (in the host document)? That's a limitation of having the trap work on a child document.

No. I can still use tab and shift+tab to navigate through other elements in the child document. I'm aware of the limitations of different contexts and that you can't trap focus in one window to another window. That's not what I'm referring to.

> Also, just to check everything, you should be using focus-trap-react v8.8.0 or greater. That's the version that added support for the focus-trap `document` option.

Just installed it a couple days ago. Using version 8.8.2.

**Comment by stefcameron** [comment]
Right, not much to go on then... Are you able to debug into focus-trap-react.js to try and see where the problem is? If your node_modules tree isn't read-only, you could at least add console statements in focus-trap-react.js and then see what parts of the code get executed at runtime and which don't. That might shed some light on things.

**Comment by LoganDark** [comment]
> Right, not much to go on then... Are you able to debug into focus-trap-react.js to try and see where the problem is? If your node_modules tree isn't read-only, you could at least add console statements in focus-trap-react.js and then see what parts of the code get executed at runtime and which don't. That might shed some light on things.

The Chrome debugger has code coverage so I can just use that. (It shows timings on the left for lines that got executed)

Could be a couple days though, I'll keep you posted. In the meantime I can provide a codesandbox that reproduces the issue?

**Comment by stefcameron** [comment]
Take your time. A repro always helps, for sure!

**Comment by LoganDark** [comment]
https://codesandbox.io/s/focus-trap-react-other-document-6mvjj

**Comment by LoganDark** [comment]
Huh. Actually it seems specifying `document` as a focusTrapOption now fixes it in the codesandbox. Weird.

**Comment by stefcameron** [comment]
I can't repro in either the native window or the popup one:

![native](https://user-images.githubusercontent.com/2855350/144723056-753f61d7-801a-4b20-9aa3-99f2d1c4e497.gif)

![window](https://user-images.githubusercontent.com/2855350/144723057-f5ef8db8-c335-4b9a-8821-49ad163f9bd3.gif)

**Comment by LoganDark** [comment]
> I can't repro in either the native window or the popup one:
> 
> ![native](https://user-images.githubusercontent.com/2855350/144723056-753f61d7-801a-4b20-9aa3-99f2d1c4e497.gif)
> 
> ![window](https://user-images.githubusercontent.com/2855350/144723057-f5ef8db8-c335-4b9a-8821-49ad163f9bd3.gif)

Yeah, that's because specifying document seems to have fixed it in the codesandbox for whatever reason.

**Comment by LoganDark** [comment]
![image](https://user-images.githubusercontent.com/4723091/144723251-3cdbe2cc-464f-4a53-8e2c-c94f2ce5c2c2.png)

Wrong document here.

The issue seems to be that the focus trap is created before `active` is first set to `true`, so `createFocusTrap` is latching onto the wrong document (because `document` is still `undefined`, because it takes a state update to stabilize). Then `doc` gets stored forever and never gets corrected.

**Comment by stefcameron** [comment]
So something like `componentDidMount()`, only call `this.setupFocusTrap()` if `this.props.active` is true, and then in `componentDidUpdate()`, check if `(!prevProps.active && this.props.active)` and call `this.setupFocusTrap()` (if it doesn't already exist).

**Comment by stefcameron** [comment]
And there's no way you can change your code to delay the inclusion of the focus trap component to when the document has stabilized? Like don't render the trap until you have the document you need...

**Comment by stefcameron** [comment]
Because this is kind of a big change to the way it's currently working.

**Comment by LoganDark** [comment]
> And there's no way you can change your code to delay the inclusion of the focus trap component to when the document has stabilized? Like don't render the trap until you have the document you need...

This... is impossible? Like, my component doesn't know where it's mounted until it's been mounted.

> Because this is kind of a big change to the way it's currently working.

It's not. It's highly unlikely people are relying on when the document field is read from the options. If anything this fixes a bug.

**Comment by stefcameron** [comment]
> > And there's no way you can change your code to delay the inclusion of the focus trap component to when the document has stabilized? Like don't render the trap until you have the document you need...
> 
> This... is impossible? Like, my component doesn't know where it's mounted until it's been mounted.

That's true. Then rendering with `<FocusTrap>` could cause the div to change, causing the ref to change, and I guess it could be a different document at that point, though that seems unlikely. Still, it wouldn't be totally "clean" to assume it's the same document.

> > Because this is kind of a big change to the way it's currently working.
> 
> It's not. It's highly unlikely people are relying on when the document field is read from the options. If anything this fixes a bug, where the focus trap reads the options before they are ready.

I'll give this some thought. It's a significant change in terms of when the trap gets created. I'll have to check focus-trap to see if creating the trap triggers any events for which you might expect a callback in one of the options, in which case, you wouldn't get the callback at the same time anymore.

If that turns out to be a problem, then another option might be to promote the `document` option to a prop (instead of nesting it in `focusTrapOptions` where there's no "notification" if it changes) and then re-create the trap if that prop changes. Then we could maintain existing behavior and only have changes affect changing the `document`.

**Comment by LoganDark** [comment]
> If that turns out to be a problem, then another option might be to promote the `document` option to a prop (instead of nesting it in `focusTrapOptions` where there's no "notification" if it changes) and then re-create the trap if that prop changes. Then we could maintain existing behavior and only have changes affect changing the `document`.

Either that or make it so that I don't have to specify the `document` field at all. You have access to the DOM node, so just get it from there maybe? The `document` field should only be necessary if I'm doing something exotic.

**Comment by stefcameron** [comment]
@LoganDark Can you check if #546 will fix your issue?

**Comment by stefcameron** [comment]
> Either that or make it so that I don't have to specify the `document` field at all. You have access to the DOM node, so just get it from there maybe? The `document` field should only be necessary if I'm doing something exotic.

I did consider this, but if you give the trap more than one container, which doc do we use (I guess they should all be from the same document, but still), and more, if the container elements are updated while the trap is active, we'd have to check for a change in the doc again and then remove/re-add all the listeners... Just seems like too risky when there seems to be a simpler solution, which I hope I have in #546.

**Comment by LoganDark** [comment]
> @LoganDark Can you check if #546 will fix your issue?

Definitely seems like it.

**Comment by stefcameron** [comment]
OK, I'll merge it and publish it. Hopefully it addresses your issue.

**Comment by stefcameron** [comment]
It'll be published imminently in 8.9.0.

**Comment by stefcameron** [comment]
@all-contributors add @LoganDark for bug
